src/routing_table.rs

```diff
@@ -0,0 +1,205 @@
+// Copyright 2020 MaidSafe.net limited.
+//
+// This SAFE Network Software is licensed to you under The General Public License (GPL), version 3.
+// Unless required by applicable law or agreed to in writing, the SAFE Network Software distributed
+// under the GPL Licence is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied. Please review the Licences for the specific language governing
+// permissions and limitations relating to use of the SAFE Network Software.
+
+//! Utilities for routing messages through the network.
+
+use crate::{
+    error::{Result, RoutingError},
+    id::{P2pNode, PublicId},
+    location::DstLocation,
+    section::{EldersInfo, SectionMap, SectionMembers},
+    xor_space::{XorName, Xorable},
+};
+use itertools::Itertools;
+
+/// Returns the delivery group size based on the section size `n`
+pub const fn delivery_group_size(n: usize) -> usize {
+    // this is an integer that is ≥ n/3
+    (n + 2) / 3
+}
+
+/// Returns a set of nodes to which a message for the given `DstLocation` could be sent
+/// onwards, sorted by priority, along with the number of targets the message should be sent to.
+/// If the total number of targets returned is larger than this number, the spare targets can
+/// be used if the message can't be delivered to some of the initial ones.
+///
+/// * If the destination is an `DstLocation::Section`:
+///     - if our section is the closest on the network (i.e. our section's prefix is a prefix of
+///       the destination), returns all other members of our section; otherwise
+///     - returns the `N/3` closest members to the target
+///
+/// * If the destination is an `DstLocation::PrefixSection`:
+///     - if the prefix is compatible with our prefix and is fully-covered by prefixes in our
+///       RT, returns all members in these prefixes except ourself; otherwise
+///     - if the prefix is compatible with our prefix and is *not* fully-covered by prefixes in
+///       our RT, returns `Err(Error::CannotRoute)`; otherwise
+///     - returns the `N/3` closest members of the RT to the lower bound of the target
+///       prefix
+///
+/// * If the destination is an individual node:
+///     - if our name *is* the destination, returns an empty set; otherwise
+///     - if the destination name is an entry in the routing table, returns it; otherwise
+///     - returns the `N/3` closest members of the RT to the target
+pub fn delivery_targets(
+    dst: &DstLocation,
+    our_id: &PublicId,
+    our_members: &SectionMembers,
+    sections: &SectionMap,
+) -> Result<(Vec<P2pNode>, usize)> {
+    if !sections.is_elder(our_id) {
```

Comment:
I thought being a non-elder, we shall not relay message?
i.e. message flow from elder to adult/infant is allowed, but the backwards is not?

Reply:
We do relay them, but only to our elders. This is because a node might receive a message addressed to elder but the node is not elder yet or not elder anymore. To prevent losing the message, we relay it to the elders to let them deal with it.